library/std/src/sys/pal/unix/kernel_copy.rs

```diff
@@ -619,7 +627,28 @@ pub(super) fn copy_regular_files(reader: RawFd, writer: RawFd, max_len: u64) ->
                 return match err.raw_os_error() {
                     // when file offset + max_length > u64::MAX
                     Some(EOVERFLOW) => CopyResult::Fallback(written),
-                    Some(ENOSYS | EXDEV | EINVAL | EPERM | EOPNOTSUPP | EBADF) if written == 0 => {
+                    Some(raw_os_error @ (ENOSYS | EXDEV | EINVAL | EPERM | EOPNOTSUPP | EBADF))
+                        if written == 0 =>
+                    {
+                        if !have_probed {
+                            let available = if matches!(raw_os_error, ENOSYS | EOPNOTSUPP | EPERM) {
+                                // EPERM can indicate seccomp filters or an
+                                // immutable file. To distinguish these
+                                // cases we probe with invalid file
+                                // descriptors which should result in EBADF
+                                // if the syscall is supported and EPERM or
+                                // ENOSYS if it's not available.
+                                //
+                                // For EOPNOTSUPP, see below. In the case of
+                                // ENOSYS, we check even though it should be
+                                // clear that there's no support.
```

Comment:
The comment about ENOSYS is not really helpful. It's meant to cover unexpected edge-cases, such as the FUSE potentially returning that error. By not having additional heuristics we don't need to worry about those things.

Reply:
I unresolved the following conversation: https://github.com/rust-lang/rust/pull/122079#discussion_r1519980059.

> It's meant to cover unexpected edge-cases, such as the FUSE potentially returning that error. By not having additional heuristics we don't need to worry about those things.

That seems unnecessary. The worst thing that can happen with a buggy FUSE driver returning ENOSYS (which it clearly shouldn't) is that we use the fallback on systems that actually support `copy_file_range`. Buggy FUSE driver in, slightly less performant Rust program out. Doesn't sound too bad to me.

If we decide to care about buggy FUSE drivers, should I make a PR for other parts of the standard library where they could cause a performance degradation by fallbacks? (linked above)

Reply:
The fallback is global process state. If for example one writes a file indexing daemon then encountering such a fuse filesystem even once would degrade performance for the whole system relying on the daemon until it is restarted.

It's simply a semantic mismatch. The static represents whether the operating systems supports the syscall. The fuse filesystem does not represent the whole system, it is one out of N things.

Reply:
> The fallback is global process state. If for example one writes a file indexing daemon then encountering such a fuse filesystem even once would degrade performance for the whole system relying on the daemon until it is restarted.
> 
> It's simply a semantic mismatch. The static represents whether the operating systems supports the syscall. The fuse filesystem does not represent the whole system, it is one out of N things.

I'll create a PR for the statx thing then I guess.

> Additionally, I take it as an example of a more general problem: Fewer edge-cases to consider. It's not specifically about fuse, though that's a salient example.

With ENOSYS, the kernel clearly communicates that the syscall does not exist. Trying to do defensive coding here sounds like the wrong thing to do here.

Reply:
#123928.

Reply:
> With ENOSYS, the kernel clearly communicates that the syscall does not exist. Trying to do defensive coding sounds like the wrong thing to do here.

It'd be great if that were the case. But error codes listed in manpages are incomplete, they're occasionally used in funky ways, there are uncommon code paths that can deviate from conventions (such as FUSE) and there even are the [occasional kernel bugs](https://github.com/rust-lang/rust/blob/31cfa4a95611b1151bcdee213d39e2edbeee6c19/library/std/src/sys/unix/thread.rs#L332-L334) were explicit API guarantees are violated. In sum there are sufficient opportunities for misbehavior to justify defensive coding.

The one case we explicitly don't handle is seccomp filters changing during process lifetime, because that would render probing futile and we'd have to pay the costs all the time instead of a single time per process.

Reply:
> there even are the [occasional kernel bugs](https://github.com/rust-lang/rust/blob/31cfa4a95611b1151bcdee213d39e2edbeee6c19/library/std/src/sys/unix/thread.rs#L332-L334) were explicit API guarantees are violated.

But we definitely do not try to preemptively avoid those.